### DISCUSSION TITLE
Question: Can kuzu embedded run in the browser via WASM?

### DISCUSSION THREAD

**Discussion by ShravanSunder** [kuzudb/kuzu#3327] [Q&A] [answered]
Can the node.js client run from a browser (esm) context (similar to duckdb).   I couldn't find information about it.  I was wondering if i could run it from cloudflare workers using httpfs and s3 as a datasource.  

from my understanding, duckdb uses a wasm import.  I see some references to wasm in your codebase, but i wasn't sure.  

thanks,

**Comment by mewim** [comment, marked as answer]
Hi @ShravanSunder,

Thanks for your interest. I think you are referring to https://github.com/duckdb/duckdb-wasm, which support running DuckDB inside browser. I started some preliminary work on getting a wasm version of Kùzu on https://github.com/kuzudb/kuzu/tree/emcc-test, but it does not quite work right now as Kùzu’s current buffer manager uses memory mapping and requires support of `MADV_DONTNEED`, which is not available for wasm right now. Later this year, we may work on an alternative version of buffer manager which does not depend on memory mapping so that wasm can be supported. I will convert this issue into a discussion and let you know when wasm support is available.

**Reply by ShravanSunder** [reply]
@mewim thanks for the response!

A secondary question.  Has there been prior art on running kuzu on lambdas?  How does the memory requirements work with httpfs?  I was also unable to find info on this while searchign the repo.   Does it scale with entities, or does httpfs use streaming from data source work through issues?

ideally i want to run the graph db in a serverless fashion and i'm wondering if thats currently possible with kuzu?

**Reply by prrao87** [reply]
Hi @ShravanSunder, it's possible to run Kùzu on Lambdas, just like you would any other embedded DB like DuckDB or LanceDB. We haven't published examples of this yet, but the memory requirements are whatever limits set by the lambdas. The typical use case for this would be when you need many relatively small, short-lived graphs (though Kùzu does scale on a single node to very large graphs of billions of nodes as well).

So, to answer your question about serverless usage, it should work and we'll be looking to get some material out on that publicly soon, but in the meantime, feel free to experiment, and reach out to us [on Discord](https://discord.com/invite/VtX2gw9Rug) where other members from the community who've worked with serverless usecases may also be able to chip in.
